File: nodes/int_nodes.py
class IntDivisibleBy_StDismas:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "value": ("INT", {"default": 0, "min": 0, "max": 2147483647, "step": 1}),
                "divisible_by": ("INT", {"default": 1, "min": 1, "max": 2147483647, "step": 1}),
            }
        }

    RETURN_TYPES = ("INT",)
    RETURN_NAMES = ("value",)
    FUNCTION = "run"
    CATEGORY = "Comfyui-StDismas/Utils/Int"

    def run(self, value, divisible_by):
        divisible_by = max(1, int(divisible_by))
        value = max(0, int(value))
        value = round(value / divisible_by) * divisible_by
        return (int(value),)


class SetDimension_StDismas:
    FORMATS = ["none", "LTXV", "WAN", "F.Klein"]

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "value": ("INT", {"default": 0, "min": 0, "max": 30720, "step": 1}),
                "format": (cls.FORMATS, {"default": "none"}),
            }
        }

    RETURN_TYPES = ("INT",)
    RETURN_NAMES = ("value",)
    FUNCTION = "run"
    CATEGORY = "Comfyui-StDismas/Utils/Int"

    @staticmethod
    def _rules(format_name):
        if format_name == "LTXV":
            return 32, 0, 0, 30720
        if format_name == "WAN":
            return 8, 0, 0, 30720
        if format_name == "F.Klein":
            return 16, 0, 0, 30720
        return 1, 0, 0, 30720

    def run(self, value, format):
        step, mod, min_v, max_v = self._rules(format)
        value = max(min_v, min(max_v, int(value)))
        value = int(round((value - mod) / step) * step + mod)
        value = max(min_v, min(max_v, value))
        return (value,)


class SetDuration_StDismas:
    FORMATS = ["none", "LTXV", "WAN"]

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "value": ("INT", {"default": 1, "min": 0, "max": 2147483647, "step": 1}),
                "format": (cls.FORMATS, {"default": "none"}),
            }
        }

    RETURN_TYPES = ("INT",)
    RETURN_NAMES = ("frame_count",)
    FUNCTION = "run"
    CATEGORY = "Comfyui-StDismas/Utils/Int"

    @staticmethod
    def _rules(format_name):
        if format_name == "LTXV":
            return 8, 1, 1, 2147483647
        if format_name == "WAN":
            return 4, 1, 1, 2147483647
        return 1, 0, 0, 2147483647

    def run(self, value, format):
        step, mod, min_v, max_v = self._rules(format)
        value = max(min_v, min(max_v, int(value)))
        if step == 1 and mod == 0:
            return (value,)
        value = int(round((value - mod) / step) * step + mod)
        value = max(min_v, min(max_v, value))
        return (value,)


NODE_CLASS_MAPPINGS = {
    "IntDivisibleBy_StDismas": IntDivisibleBy_StDismas,
    "SetDimension_StDismas": SetDimension_StDismas,
    "SetDuration_StDismas": SetDuration_StDismas,
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "IntDivisibleBy_StDismas": "Int Divisible By (StDismas)",
    "SetDimension_StDismas": "Set Dimension (StDismas)",
    "SetDuration_StDismas": "Set Duration (StDismas)",
}

File: web/int_nodes.js
import { app } from "../../scripts/app.js";

function clamp(v, min, max) {
  if (min != null && v < min) v = min;
  if (max != null && v > max) v = max;
  return v;
}

function snapToRule(value, options = {}) {
  let v = Number(value);
  if (!Number.isFinite(v)) v = Number(options.default ?? 0);

  const min = options.min;
  const max = options.max;
  const step = Math.max(1, Number(options.step ?? 1));
  const mod = Number(options.mod ?? 0);

  v = clamp(v, min, max);
  v = Math.round((v - mod) / step) * step + mod;
  v = clamp(v, min, max);
  return Math.trunc(v);
}

function updateWidgetValue(node, widget, newValue) {
  const snapped = snapToRule(newValue, widget.options || {});
  widget.value = snapped;
  if (widget.options?.property && node.properties && widget.options.property in node.properties) {
    node.setProperty(widget.options.property, snapped);
  }
  node.setDirtyCanvas?.(true, true);
  node.graph?.setDirtyCanvas(true, true);
}

function patchValueWidget(node, valueWidget) {
  if (!valueWidget || valueWidget.__stdismasPatched) return;
  valueWidget.__stdismasPatched = true;

  const originalCallback = valueWidget.callback?.bind(valueWidget);
  valueWidget.callback = function (v, ...args) {
    const snapped = snapToRule(v, this.options || {});
    this.value = snapped;
    if (originalCallback) {
      return originalCallback(snapped, ...args);
    }
    node.graph?.setDirtyCanvas(true, true);
  };

  const originalMouse = valueWidget.mouse?.bind(valueWidget);
  valueWidget.mouse = function (event, pos, n) {
    const result = originalMouse ? originalMouse(event, pos, n) : undefined;
    const snapped = snapToRule(this.value, this.options || {});
    if (this.value !== snapped) {
      this.value = snapped;
      node.graph?.setDirtyCanvas(true, true);
    }
    return result;
  };
}

function findWidget(node, name) {
  return node.widgets?.find((w) => w.name === name);
}

function dimensionRule(format) {
  switch (format) {
    case "LTXV":
      return { min: 0, max: 30720, step: 32, mod: 0 };
    case "WAN":
      return { min: 0, max: 30720, step: 8, mod: 0 };
    case "F.Klein":
      return { min: 0, max: 30720, step: 16, mod: 0 };
    default:
      return { min: 0, max: 30720, step: 1, mod: 0 };
  }
}

function durationRule(format) {
  switch (format) {
    case "LTXV":
      return { min: 1, max: 2147483647, step: 8, mod: 1 };
    case "WAN":
      return { min: 1, max: 2147483647, step: 4, mod: 1 };
    default:
      return { min: 0, max: 2147483647, step: 1, mod: 0 };
  }
}

function divisibleRule(divisibleBy) {
  const d = Math.max(1, Math.trunc(Number(divisibleBy) || 1));
  return { min: 0, max: 2147483647, step: d, mod: 0 };
}

function applyRule(node, valueWidget, rule) {
  if (!valueWidget) return;
  valueWidget.options = valueWidget.options || {};
  valueWidget.options.min = rule.min;
  valueWidget.options.max = rule.max;
  valueWidget.options.step = rule.step;
  valueWidget.options.mod = rule.mod;
  updateWidgetValue(node, valueWidget, valueWidget.value);
}

function hookSourceWidget(node, sourceWidget, apply) {
  if (!sourceWidget || sourceWidget.__stdismasRuleHooked) return;
  sourceWidget.__stdismasRuleHooked = true;

  const originalCallback = sourceWidget.callback?.bind(sourceWidget);
  sourceWidget.callback = function (v, ...args) {
    const result = originalCallback ? originalCallback(v, ...args) : undefined;
    setTimeout(() => apply(), 0);
    return result;
  };

  const originalMouse = sourceWidget.mouse?.bind(sourceWidget);
  sourceWidget.mouse = function (event, pos, n) {
    const result = originalMouse ? originalMouse(event, pos, n) : undefined;
    setTimeout(() => apply(), 0);
    return result;
  };
}

function setupReactiveRule(node, kind) {
  const valueWidget = findWidget(node, "value");
  patchValueWidget(node, valueWidget);
  if (!valueWidget) return;

  let sourceWidget;
  let getRule;

  if (kind === "dimension") {
    sourceWidget = findWidget(node, "format");
    getRule = () => dimensionRule(sourceWidget?.value);
  } else if (kind === "duration") {
    sourceWidget = findWidget(node, "format");
    getRule = () => durationRule(sourceWidget?.value);
  } else if (kind === "divisible") {
    sourceWidget = findWidget(node, "divisible_by");
    patchValueWidget(node, sourceWidget);
    getRule = () => divisibleRule(sourceWidget?.value);
  }

  if (!sourceWidget || !getRule) return;

  const apply = () => applyRule(node, valueWidget, getRule());
  hookSourceWidget(node, sourceWidget, apply);
  apply();
}

app.registerExtension({
  name: "stdismas.int_nodes_builtin",
  nodeCreated(node) {
    if (node.comfyClass === "SetDimension_StDismas") {
      setupReactiveRule(node, "dimension");
    } else if (node.comfyClass === "SetDuration_StDismas") {
      setupReactiveRule(node, "duration");
    } else if (node.comfyClass === "IntDivisibleBy_StDismas") {
      setupReactiveRule(node, "divisible");
    }
  },
});
